algosdk/box_reference.py

```diff
@@ -0,0 +1,92 @@
+from collections import OrderedDict
+from typing import List, Tuple
+
+from algosdk import error
+
+
+class BoxReference:
+    """
+    Represents a box reference with a foreign app index and the box name.
+
+    Args:
+        app_index (int): index of the application in the foreign app array
+        name (bytes): key for the box in bytes
+    """
+
+    def __init__(self, app_index: int, name: bytes):
+        self.app_index = app_index
+        self.name = name
+
+    @staticmethod
+    def translate_box_reference(
+        ref: Tuple[int, bytes],
+        foreign_apps: List[int],
+        this_app_id: int,
+    ) -> "BoxReference":
+        # Try coercing reference id and name.
+        from algosdk.future.transaction import ApplicationCallTxn
+
+        ref_id, ref_name = int(ref[0]), ApplicationCallTxn.as_bytes(ref[1])
+        index = 0
+        try:
+            # Foreign apps start from index 1; index 0 is its own app ID.
+            index = foreign_apps.index(ref_id) + 1
+        except ValueError:
+            # Check if the app referenced is itself after checking the
+            # foreign apps array (in case its own app id is in its own
+            # foreign apps array).
+            if ref_id == 0 or ref_id == this_app_id:
+                pass
+            else:
+                raise error.InvalidForeignAppIdError(
+                    f"Box ref with appId {ref_id} not in foreign-apps"
+                )
+        return BoxReference(index, ref_name)
+
+    @staticmethod
+    def translate_box_references(
+        references: List[Tuple[int, bytes]],
+        foreign_apps: List[int],
+        this_app_id: int,
+    ) -> List["BoxReference"]:
+        """
+        Translates a list of tuples with app IDs and names into an array of
+        BoxReferences with foreign indices.
+
+        Args:
+            references (list[(int, bytes)]): list of tuples specifying app id
+                and key for boxes the app may access
+            foreign_apps (list[int]): list of other applications in appl call
+            this_app_id (int): app ID of the box being references
+        """
+        if not references:
+            return []
+
+        return [
+            BoxReference.translate_box_reference(
+                ref, foreign_apps, this_app_id
+            )
+            for ref in references
+        ]
+
+    def dictify(self):
+        d = dict()
```

Comment:
if truthiness of `app_index` and `name` are enforced in `__init__()`, there will be no need to have this checks.

Reply:
I think this has to do with the msgpack encoding as we omit empty/zeroed fields when sending them over the wire. JJ mentioned that Box names and app ids can be 0's (which interestingly means that you can specify a box that will have just empty fields in the msgpack 😱 )